receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/cpu_utilization_calculator.go

```diff
@@ -0,0 +1,102 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+
+	"github.com/shirou/gopsutil/v3/cpu"
+	"go.opentelemetry.io/collector/model/pdata"
+)
+
+// CPUUtilization stores the utilization percents [0-1] for the different cpu states
+type CPUUtilization struct {
+	CPU     string
+	User    float64
+	System  float64
+	Idle    float64
+	Nice    float64
+	Iowait  float64
+	Irq     float64
+	Softirq float64
+	Steal   float64
+}
+
+// CPUUtilizationCalculator calculates the cpu utilization percents for the different cpu states
+// It requires 2 []cpu.TimesStat and spend time to be able to calculate the difference
+type CPUUtilizationCalculator struct {
+	previousTime     pdata.Timestamp
+	previousCPUTimes []cpu.TimesStat
+}
+
+// Calculate calculates the cpu utilization for the different cpu states comparing previously
+// stored []cpu.TimesStat and time.Time and current []cpu.TimesStat and current time.Time
+// If no previous data is stored it will return empty slice of CPUUtilization and no error
+func (c *CPUUtilizationCalculator) Calculate(now pdata.Timestamp, cpuTimes []cpu.TimesStat) ([]CPUUtilization, error) {
+	var utilizationByCPU []CPUUtilization
+	elapsedSeconds := now.AsTime().Sub(c.previousTime.AsTime()).Seconds()
+
+	if c.previousCPUTimes != nil {
+		for _, previousCPUTime := range c.previousCPUTimes {
+			currentCPUTime, err := cpuTimeByCPU(previousCPUTime.CPU, cpuTimes)
+			if err != nil {
+				return []CPUUtilization{}, err
+			}
+			utilization, err := cpuUtilization(previousCPUTime, currentCPUTime, elapsedSeconds)
+			if err != nil {
+				return []CPUUtilization{}, err
+			}
+			utilizationByCPU = append(utilizationByCPU, utilization)
+		}
+	}
+	c.previousCPUTimes = cpuTimes
+	c.previousTime = now
+
+	return utilizationByCPU, nil
+}
+
+// cpuUtilization calculates the difference between 2 cpu.TimesStat using spent time between them
+// If no time was spent between TimesStat an error will be returned
+// If TimesStats do not belog to same CPU an error will be returned
+func cpuUtilization(timeStart cpu.TimesStat, timeEnd cpu.TimesStat, elapsedSeconds float64) (CPUUtilization, error) {
+	if elapsedSeconds <= 0 {
+		return CPUUtilization{}, fmt.Errorf("invalid elapsedSeconds: %f", elapsedSeconds)
```

Comment:
Creating errors "from scratch" with `fmt.Errorf` makes them difficult to compare later, as there is not a base error anywhere.

This can be solved by creating a package-level error like

```go
var ErrInvalidElapsed = errors.New("invalid elapsed seconds")
```

This can later be wrapped:
```suggestion
		return CPUUtilization{}, fmt.Errorf("%f: %w", elapsedSeconds, ErrInvalidElapsed)
```

By doing this, `cpuUtilization` callers can use `errors.Is(err, ErrInvalidElapsed)` on the returned error.